modules/programs/fish.nix

```diff
@@ -439,11 +439,12 @@ in {
         # add completions generated by Home Manager to $fish_complete_path
         begin
           set -l joined (string join " " $fish_complete_path)
-          set -l prev_joined (string replace --regex "[^\s]*generated_completions.*" "" $joined)
+          set -l prev_joined (string replace --regex "[^\s]*home-manager-path/share/fish/vendor_completions\.d" "" $joined)
```

Comment:
Is this line actually needed? I don't quite understand what this `prev_joined` is for but I think it's just to make sure that path with `*generated_completions.*` is in front?

I think this PR should do is to handle the vendor completions generated by NixOS system configuration which I believe is `/etc/profiles/per-user/<username>/share/fish/vendor_completions.d`?

Reply:
Hello :)

This begin...end section constructs the fish shell variable `fish_complete_path`. Path variables in fish are handled like lists; however, there isn't a friendly API to 
- move an element to the front if it's not in the right position
- do nothing if the element exists
- remove an element matching a regex (not as I know without using a loop)
This is most likely done to ensure idempotency.

The solution to this (done by contributor who wrote this section before me) joined the list into a string, removes the generated completion paths using regex, and then update the `fish_complete_path`. The use of regex here is notable, it is here to remove all generated completion, because they are prefixed with nix store's cryptographic hash.

For the above reasons, I don't think we can remove this line. This might result in nix sourcing multiple times the same path. 

I didn't source what's in `/etc/profiles/per-user/<username>/share/fish/vendor_completions.d`, because this would be an invalid path if someone were to use `home-manager` as a standalone module (or with `home-manager.useUserPackages = false;`)
`"${config.home.path}/share/fish/vendor_completions.d"` refers to the vendor_completions.d aggregated in home-manager's linked path.

Have a nice day :)

Reply:
Thank you for the explanation, what I meant by handling the vendor completion by system module is when user has `programs.fish.vendor.completions.enable = true;` in the configuration.nix file. I believe this will cause duplicate vendor completions? It's not deal breaker but sort of innefficient I think.

Reply:
> when user has programs.fish.vendor.completions.enable = true; in the configuration.nix file. I believe this will cause duplicate vendor completions?

Interesting, I didn't think of this.
What is proposed in this PR will indeed created duplicated vendor completions. Is there a way to detect whether a configuration is being used as a NixOS module or not, so that we can decide which path to source programmatically ?
`/etc/profiles/per-user/<username>/share/fish/vendor_completions.d` won't exist if the user doesn't enable NixOS managed `fish`, or if home-manager is was installed as standalone.
Maybe in `config.fish` script we can detect whether `/share/fish/vendor_completions.d` in `/etc` exists and then source that conditionally ?

Reply:
I don't know for sure the best way to do this, but maybe using `fish` `string replace --regex` to filter them out? Regex something like `*/per-user/\.+/share/fish/vendor_completions.d`? But I think this should be a module option that user can switch on/off in case this break anything? Maybe @rycee or other maintainers can provide better idea on this.